libs/utils/automated-aria.js

```diff
@@ -0,0 +1,263 @@
+};
+
+// Helper to extract prefix or suffix from class name parts
+const getPatternPart = (parts, usePrefix) => (usePrefix ? parts[0] : parts[parts.length - 1]);
+
+// Helper to check if two divs share a common class pattern
+const divsSharePattern = (div1, div2, hasCommonPrefix, hasCommonSuffix) => {
+  const div1Classes = Array.from(div1.classList);
+  const div2Classes = Array.from(div2.classList);
+  return div1Classes.some((cn1) => div2Classes.some((cn2) => {
+    const parts1 = cn1.split('-');
+    const parts2 = cn2.split('-');
+    const hasPrefixMatch = hasCommonPrefix && parts1[0] === parts2[0];
+    // eslint-disable-next-line max-len
+    const hasSuffixMatch = hasCommonSuffix && parts1[parts1.length - 1] === parts2[parts2.length - 1];
+    return parts1.length > 1 && parts2.length > 1 && (hasPrefixMatch || hasSuffixMatch);
+  }));
+};
+
+// Helper to check if divs have a common pattern (prefix or suffix)
+const hasCommonPattern = (divs, usePrefix) => {
+  const checkFn = (parts) => getPatternPart(parts, usePrefix);
+  const classNames = divs.flatMap((div) => Array.from(div.classList));
+  const patterns = new Set(classNames
+    .map((cn) => cn.split('-'))
+    .filter((parts) => parts.length > 1)
+    .map(checkFn));
+  return patterns.size > 0 && divs.some((div1) => divs
+    .some((div2) => div1 !== div2 && divsSharePattern(div1, div2, usePrefix, !usePrefix)));
+};
+
+// Helper to process nested containers recursively
+const processNestedContainers = (patternDivs, findBlockContainersFn) => {
+  const deepestContainers = [];
+  patternDivs.forEach((div) => {
+    const nested = findBlockContainersFn(div);
+    if (nested.length === 1 && nested[0] === div) {
+      deepestContainers.push(div);
+    } else if (nested.length > 1) {
+      deepestContainers.push(...nested);
+      deepestContainers.push(div); // Add the parent last
+    }
+  });
+  return deepestContainers;
+};
+
+// Identifies container divs within a block that share a common class pattern
+// and appear multiple times at the same DOM level.
+// Containers must be siblings with a shared prefix (e.g., 'card-')
+// or suffix (e.g., '-item') and occur more than once.
+export const findBlockContainers = (container) => {
+  // Get all direct child divs of the container
+  const directDivs = Array.from(container.children).filter((el) => el.tagName === 'DIV');
+  if (!directDivs.length) return [container];
+
+  // Filter divs that share a common pattern (prefix or suffix)
+  const patternDivs = directDivs.filter((div) => directDivs.some((otherDiv) => div !== otherDiv && (
+    hasCommonPattern([div, otherDiv], true) // has common prefix
+        || hasCommonPattern([div, otherDiv], false) // has common suffix
+  )));
+
+  // Require at least two matching siblings to qualify as containers
+  if (patternDivs.length > 1) {
+    const deepestContainers = processNestedContainers(patternDivs, findBlockContainers);
```

Comment:
Do you need to pass in `findBlockContainers` ? Can't you just call the function

Reply:
Yes, unfortunately it's necessary. The issue is findDeepestRepeatingContainers is called by findBlockContainers and must therefore appear before. But findDeepestRepeatingContainers also calls findBlockContainers since it's recursive, and findBlockContainers appears after.